qiskit/primitives/containers/bindings_array.py

```diff
@@ -169,6 +169,70 @@ def vals(self) -> list[np.ndarray]:
         """The non-keyword values of this array."""
         return self._vals
 
+    def as_array(self, parameters: Iterable[Parameter | str] | None = None) -> np.ndarray:
+        """Return the contents of this bindings array as a single NumPy array.
+
+        As with each :attr:`~vals` and :attr:`~kwvals` array, the parameters are indexed along the
+        last dimension.
+
+        The order of the :attr:`~vals` portion of this bindings array is always preserved, and
+        always comes first in the returned array, irrespective of whether ``parameters`` are
+        provided.
+
+        If ``parameters`` are provided, then they determine the order of any :attr:`~kwvals`
+        present in this bindings array. If :attr:`~vals` are present in addition to :attr:`~kwvals`,
+        then it is up to the user to ensure that their provided ``parameters`` account for this.
+
+        Parameters:
+            parameters: Optional parameters that determine the order of the output.
+
+        Returns:
+            This bindings array as a single NumPy array.
+
+        Raises:
+            RuntimeError: If these bindings contain multiple dtypes.
+            KeyError: If ``parameters`` are provided that are not a superset of those in this
+                bindings array.
+        """
+        dtypes = {arr.dtype for arr in self.vals}
+        dtypes.update(arr.dtype for arr in self.kwvals.values())
+        if len(dtypes) > 1:
+            raise RuntimeError(f"Multiple dtypes ({dtypes}) were found.")
```

Comment:
It might be good to validate dtype in `validate` method instead of here.

Reply:
Support for multiple dtypes within a BindingsArray instance is intentional, reflecting the idea that different `Parameter`s might want different types. So I think the validation needs to happen here.

BTW, if we only supported one dtype for a whole bindings array, we could simplify the data model and just have one array for `vals` and one array for `kwvals`.

Reply:
I see. Multiple dtypes does not cause a problem unless invoking `as_array`. It makes sense to put dtype validation here.